File: nop/ts/admin/src/protocol/system.ts
// This file is part of the product NoPressure.
// SPDX-FileCopyrightText: 2025-2026 Zivatar Limited
// SPDX-License-Identifier: AGPL-3.0-or-later
// The code and documentation in this repository is licensed under the GNU Affero General Public License v3.0 or later (AGPL-3.0-or-later). See LICENSE.

import { WireReader, WireWriter } from "./wire";

export const SYSTEM_DOMAIN_ID = 0;
export const SYSTEM_ACTION_PING = 1;
export const SYSTEM_ACTION_PONG = 2;
export const SYSTEM_ACTION_PONG_ERROR = 3;
export const SYSTEM_ACTION_LOGGING_GET = 4;
export const SYSTEM_ACTION_LOGGING_GET_OK = 5;
export const SYSTEM_ACTION_LOGGING_GET_ERR = 6;
export const SYSTEM_ACTION_LOGGING_SET = 7;
export const SYSTEM_ACTION_LOGGING_SET_OK = 8;
export const SYSTEM_ACTION_LOGGING_SET_ERR = 9;
export const SYSTEM_ACTION_LOGGING_CLEAR = 10;
export const SYSTEM_ACTION_LOGGING_CLEAR_OK = 11;
export const SYSTEM_ACTION_LOGGING_CLEAR_ERR = 12;

export interface PingRequest {
  versionMajor: number;
  versionMinor: number;
  versionPatch: number;
}

export interface LoggingConfigRequest {}

export interface ClearLogsRequest {}

export interface LoggingSetRequest {
  rotationMaxSizeMb: number;
  rotationMaxFiles: number;
}

export interface LoggingConfigResponse {
  level: string;
  rotation_max_size_mb: number;
  rotation_max_files: number;
  run_mode: string;
  file_logging_active: boolean;
}

export interface ClearLogsResponse {
  message: string;
  deleted_files: number;
  deleted_bytes: number;
}

export interface MessageResponse {
  message: string;
}

export function encodePingRequest(payload: PingRequest): Uint8Array {
  const writer = new WireWriter();
  writer.writeU16(payload.versionMajor);
  writer.writeU16(payload.versionMinor);
  writer.writeU16(payload.versionPatch);
  return writer.toUint8Array();
}

export function encodeLoggingGetRequest(_payload: LoggingConfigRequest): Uint8Array {
  return new Uint8Array(0);
}

export function encodeLoggingSetRequest(payload: LoggingSetRequest): Uint8Array {
  const writer = new WireWriter();
  writer.writeU64(payload.rotationMaxSizeMb);
  writer.writeU32(payload.rotationMaxFiles);
  return writer.toUint8Array();
}

export function encodeLoggingClearRequest(_payload: ClearLogsRequest): Uint8Array {
  return new Uint8Array(0);
}

export function decodeLoggingConfigResponse(bytes: Uint8Array): LoggingConfigResponse {
  const reader = new WireReader(bytes);
  const level = reader.readString();
  const rotationMaxSizeMb = reader.readU64Number();
  const rotationMaxFiles = reader.readU32();
  const runMode = reader.readString();
  const fileLoggingActive = reader.readBool();
  reader.ensureFullyConsumed();
  return {
    level,
    rotation_max_size_mb: rotationMaxSizeMb,
    rotation_max_files: rotationMaxFiles,
    run_mode: runMode,
    file_logging_active: fileLoggingActive,
  };
}

export function decodeClearLogsResponse(bytes: Uint8Array): ClearLogsResponse {
  const reader = new WireReader(bytes);
  const message = reader.readString();
  const deletedFiles = reader.readU64Number();
  const deletedBytes = reader.readU64Number();
  reader.ensureFullyConsumed();
  return {
    message,
    deleted_files: deletedFiles,
    deleted_bytes: deletedBytes,
  };
}

export function decodeMessageResponse(bytes: Uint8Array): MessageResponse {
  const reader = new WireReader(bytes);
  const message = reader.readString();
  reader.ensureFullyConsumed();
  return { message };
}

File: nop/crates/nop-management-contract/src/system.rs
// This file is part of the product NoPressure.
// SPDX-FileCopyrightText: 2025-2026 Zivatar Limited
// SPDX-License-Identifier: AGPL-3.0-or-later
// The code and documentation in this repository is licensed under the GNU Affero General Public License v3.0 or later (AGPL-3.0-or-later). See LICENSE.

use crate::wire::{WireDecode, WireEncode, WireReader, WireResult, WireWriter};
use serde::{Deserialize, Serialize};

pub const SYSTEM_DOMAIN_ID: u32 = 0;
pub const SYSTEM_ACTION_PING: u32 = 1;
pub const SYSTEM_ACTION_PONG: u32 = 2;
pub const SYSTEM_ACTION_PONG_ERROR: u32 = 3;
pub const SYSTEM_ACTION_LOGGING_GET: u32 = 4;
pub const SYSTEM_ACTION_LOGGING_GET_OK: u32 = 5;
pub const SYSTEM_ACTION_LOGGING_GET_ERR: u32 = 6;
pub const SYSTEM_ACTION_LOGGING_SET: u32 = 7;
pub const SYSTEM_ACTION_LOGGING_SET_OK: u32 = 8;
pub const SYSTEM_ACTION_LOGGING_SET_ERR: u32 = 9;
pub const SYSTEM_ACTION_LOGGING_CLEAR: u32 = 10;
pub const SYSTEM_ACTION_LOGGING_CLEAR_OK: u32 = 11;
pub const SYSTEM_ACTION_LOGGING_CLEAR_ERR: u32 = 12;

#[derive(Debug, Clone)]
pub enum SystemCommand {
    Ping(PingRequest),
    GetLoggingConfig(GetLoggingConfigRequest),
    SetLoggingConfig(SetLoggingConfigRequest),
    ClearLogs(ClearLogsRequest),
}

impl SystemCommand {
    pub fn action_id(&self) -> u32 {
        match self {
            SystemCommand::Ping(_) => SYSTEM_ACTION_PING,
            SystemCommand::GetLoggingConfig(_) => SYSTEM_ACTION_LOGGING_GET,
            SystemCommand::SetLoggingConfig(_) => SYSTEM_ACTION_LOGGING_SET,
            SystemCommand::ClearLogs(_) => SYSTEM_ACTION_LOGGING_CLEAR,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PingRequest {
    pub version_major: u16,
    pub version_minor: u16,
    pub version_patch: u16,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PongResponse {
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PongErrorResponse {
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetLoggingConfigRequest {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SetLoggingConfigRequest {
    pub rotation_max_size_mb: u64,
    pub rotation_max_files: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClearLogsRequest {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoggingConfigResponse {
    pub level: String,
    pub rotation_max_size_mb: u64,
    pub rotation_max_files: u32,
    pub run_mode: String,
    pub file_logging_active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClearLogsResponse {
    pub message: String,
    pub deleted_files: u64,
    pub deleted_bytes: u64,
}

impl WireEncode for PingRequest {
    fn encode(&self, writer: &mut WireWriter) -> WireResult<()> {
        writer.write_u16(self.version_major);
        writer.write_u16(self.version_minor);
        writer.write_u16(self.version_patch);
        Ok(())
    }
}

impl WireDecode for PingRequest {
    fn decode(reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {
            version_major: reader.read_u16()?,
            version_minor: reader.read_u16()?,
            version_patch: reader.read_u16()?,
        })
    }
}

impl WireEncode for GetLoggingConfigRequest {
    fn encode(&self, _writer: &mut WireWriter) -> WireResult<()> {
        Ok(())
    }
}

impl WireDecode for GetLoggingConfigRequest {
    fn decode(_reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {})
    }
}

impl WireEncode for SetLoggingConfigRequest {
    fn encode(&self, writer: &mut WireWriter) -> WireResult<()> {
        writer.write_u64(self.rotation_max_size_mb);
        writer.write_u32(self.rotation_max_files);
        Ok(())
    }
}

impl WireDecode for SetLoggingConfigRequest {
    fn decode(reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {
            rotation_max_size_mb: reader.read_u64()?,
            rotation_max_files: reader.read_u32()?,
        })
    }
}

impl WireEncode for ClearLogsRequest {
    fn encode(&self, _writer: &mut WireWriter) -> WireResult<()> {
        Ok(())
    }
}

impl WireDecode for ClearLogsRequest {
    fn decode(_reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {})
    }
}

impl WireEncode for LoggingConfigResponse {
    fn encode(&self, writer: &mut WireWriter) -> WireResult<()> {
        writer.write_string(&self.level)?;
        writer.write_u64(self.rotation_max_size_mb);
        writer.write_u32(self.rotation_max_files);
        writer.write_string(&self.run_mode)?;
        writer.write_bool(self.file_logging_active);
        Ok(())
    }
}

impl WireDecode for LoggingConfigResponse {
    fn decode(reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {
            level: reader.read_string()?,
            rotation_max_size_mb: reader.read_u64()?,
            rotation_max_files: reader.read_u32()?,
            run_mode: reader.read_string()?,
            file_logging_active: reader.read_bool()?,
        })
    }
}

impl WireEncode for ClearLogsResponse {
    fn encode(&self, writer: &mut WireWriter) -> WireResult<()> {
        writer.write_string(&self.message)?;
        writer.write_u64(self.deleted_files);
        writer.write_u64(self.deleted_bytes);
        Ok(())
    }
}

impl WireDecode for ClearLogsResponse {
    fn decode(reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {
            message: reader.read_string()?,
            deleted_files: reader.read_u64()?,
            deleted_bytes: reader.read_u64()?,
        })
    }
}

impl WireEncode for PongResponse {
    fn encode(&self, writer: &mut WireWriter) -> WireResult<()> {
        writer.write_string(&self.message)
    }
}

impl WireDecode for PongResponse {
    fn decode(reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {
            message: reader.read_string()?,
        })
    }
}

impl WireEncode for PongErrorResponse {
    fn encode(&self, writer: &mut WireWriter) -> WireResult<()> {
        writer.write_string(&self.message)
    }
}

impl WireDecode for PongErrorResponse {
    fn decode(reader: &mut WireReader) -> WireResult<Self> {
        Ok(Self {
            message: reader.read_string()?,
        })
    }
}
